jupyter_ydoc/ydoc.py

```diff
@@ -80,20 +85,56 @@ def __init__(self, *args, **kwargs):
         self._ycells = self._ydoc.get_array("cells")
         self._ymetadata = self._ydoc.get_map("metadata")
 
-    @property
-    def source(self):
+    def get_cell(self, index: int) -> Dict[str, Any]:
+        meta = self._ymeta.to_json()
+        cell = self._ycells[index].to_json()
+        cast_all(cell, float, int)
+        if "id" in cell and meta["nbformat"] == 4 and meta["nbformat_minor"] <= 4:
+            # strip cell IDs if we have notebook format 4.0-4.4
+            del cell["id"]
+        if cell["cell_type"] in ["raw", "markdown"] and not cell["attachments"]:
+            del cell["attachments"]
+        return cell
+
+    def set_cell(self, index: int, value: Dict[str, Any], txn=None) -> None:
+        cell = copy.deepcopy(value)
+        if "id" not in cell:
+            cell["id"] = str(uuid4())
+        cell_type = cell["cell_type"]
+        cell["source"] = Y.YText(cell["source"])
+        cell["metadata"] = Y.YMap(cell.get("metadata", {}))
+        if cell_type in ("raw", "markdown"):
+            cell["attachments"] = Y.YMap(cell.get("attachments", {}))
+        elif cell_type == "code":
+            cell["outputs"] = Y.YArray(cell.get("outputs", []))
+        if len(self._ycells) < index:
+            self._ycells.delete()
+        ycell = Y.YMap(cell)
+        if txn is None:
+            with self._ydoc.begin_transaction() as txn:
+                self._put_cell(index, ycell, txn)
+        else:
+            self._put_cell(index, ycell, txn)
+
+    def _put_cell(self, index: int, ycell: Y.YMap, txn):
+        if len(self._ycells) >= index + 1:
+            self._ycells.delete(txn, index)
+        self._ycells.insert(txn, index, ycell)
```

Comment:
@Waidhoferj I'm not sure to understand the [error message](https://github.com/jupyter-server/jupyter_ydoc/runs/7270736260?check_suite_focus=true):
```
y_py.MultipleIntegrationError: Cannot integrate a nested Ypy object because is already integrated into a YDoc: {'source': YText(print('Hello, World!')), 'execution_count': None, 'metadata': YMap({}), 'outputs': YArray([]), 'id': None, 'cell_type': 'code'}
```
The [YMap](https://github.com/jupyter-server/jupyter_ydoc/pull/35/files#diff-e32c621500dede398406ab59b1fe0cf6460dba72f659e4f2809bcd85b7915928R112) I'm trying to insert into a `YArray` is not part of a `YDoc`.

Reply:
My bad, I think it was on my side.

Reply:
Actually I think there is a bug in ypy. It doesn't happen in the CI, but locally I have the error above at [this line](https://github.com/jupyter-server/jupyter_ydoc/pull/35/files#diff-e32c621500dede398406ab59b1fe0cf6460dba72f659e4f2809bcd85b7915928R128).

Reply:
@Waidhoferj ypy HEAD doesn't seem to fix this one.

Reply:
Working on it

Reply:
@davidbrochart Having trouble reproducing the issue. I've run a couple tests pushing preliminary Y elements into an array.
```python
def test_multiple_integration():
    d = Y.YDoc()
    wrapper = d.get_array("wrapper")
    inner = Y.YMap({"foo": "bar"})
    with d.begin_transaction() as txn:
        wrapper.append(txn, inner) # I'd expect an error here
```

Tried to set up the repository, but I'm having dependency issues related to ypy-websockets.

Reply:
Definitely, just release [0.5.3](https://pypi.org/project/y-py/)